packages/zoltan2/core/src/algorithms/color/Zoltan2_AlgHybrid2GL.hpp

```diff
@@ -0,0 +1,1532 @@
+      int rounds = 1;
+      for(int i = 0; i < comm->getSize(); i++){
+        if(adjsendcounts[i]*sizeof(gno_t)/ INT_MAX > (size_t)rounds){
+          rounds = (adjsendcounts[i]*sizeof(gno_t)/INT_MAX)+1;
+        }
+      }
+
+      //see what the max number of rounds will be globally
+      int max_rounds = 0;
+      Teuchos::reduceAll<int>(*comm, Teuchos::REDUCE_MAX, 1, &rounds, &max_rounds);
+      
+      if(verbose) std::cout<<comm->getRank()<<": building per_proc sums\n";
+      //compute send and receive schedules to and from each process
+      std::vector<std::vector<uint64_t>> per_proc_round_adj_sums(max_rounds+1,std::vector<uint64_t>(comm->getSize(),0));
+      std::vector<std::vector<uint64_t>> per_proc_round_vtx_sums(max_rounds+1,std::vector<uint64_t>(comm->getSize(),0));
+
+      if(verbose) std::cout<<comm->getRank()<<": filling per_proc sums\n";
+      //fill out the send schedules
+      for(int proc_to_send = 0; proc_to_send < comm->getSize(); proc_to_send++){
+        int curr_round = 0;
+        for(int j = sdispls[proc_to_send]; j < sdispls[proc_to_send+1]; j++){
+          if((per_proc_round_adj_sums[curr_round][proc_to_send] + recvDegrees[j])*sizeof(gno_t) > INT_MAX){
+            curr_round++;
+          }
+          per_proc_round_adj_sums[curr_round][proc_to_send] += recvDegrees[j];
+          per_proc_round_vtx_sums[curr_round][proc_to_send]++;
+        }
+      }
+      
+      if(verbose) std::cout<<comm->getRank()<<": building recv GID schedule\n";
+      
+      //A 3D vector to hold the GIDs so we can see how this process will receive
+      //the vertices in each round, from each process. This way we can reorder things
+      //locally so that the data we receive will automatically construct the CSR correctly
+      //without any other computation. We do the reordering before we start receiving.
+      std::vector<std::vector<std::vector<gno_t>>> recv_GID_per_proc_per_round(
+		      max_rounds+1,std::vector<std::vector<gno_t>>(
+			      comm->getSize(),std::vector<gno_t>(0)));
+      for(int i = 0; i < max_rounds; i++){
+        for(int j = 0; j < comm->getSize(); j++){
+	  recv_GID_per_proc_per_round[i][j] = std::vector<gno_t>(sendcounts[j],0);
+	}
+      }
+
+      if(verbose) std::cout<<comm->getRank()<<": filling out recv GID schedule\n";
+      for(int i = 0; i < comm->getSize(); i++){
+        int curr_round = 0;
+        size_t curr_idx = 0;
+        for(int j = sdispls[i]; j < sdispls[i+1]; j++){
+          if(curr_idx > per_proc_round_vtx_sums[curr_round][i]){
+            curr_round++;
+            curr_idx = 0;
+          }
+          recv_GID_per_proc_per_round[curr_round][i][curr_idx++] = j;
+        }
+      }
+      
+      if(verbose) std::cout<<comm->getRank()<<": reordering gids and degrees in the order they'll be received\n";
+      //reorder the GIDs and degrees locally so that the data received on this process
+      //just happens to be the correct CSR.
```

Comment:
The reason for reordering is subtle, and thus somewhat difficult for me to explain. The reordering makes the LIDs correspond to the order in which we receive information for the corresponding GIDs (the first GID info we receive is LID n_local, the second is n_local+1, etc.) This way, when we receive the adjacency information, it is in the correct order, and can safely be indexed by offsets without any modification. If I don't reorder the LIDs, I'd need to reorder the adjacencies, which I think would be much more memory-intensive, if not just messier. 
Because the original CSR doesn't have any ghost-offset information, this reorganization doesn't incur any extra cost at this point (we don't need to reorganize the original CSR); we just have to make sure the ghost LIDs are changed universally.